### DISCUSSION TITLE
Incorporating Variable Envelope Return Paths

### DISCUSSION THREAD

**Discussion by jmdunsing** [postalserver/postal#2826] [Development discussions]
I used Postal for a while now to assist with an email marketing platform that has recently seen some growth. When I began, I did not realize that postal was meant to be more of a transactional email service rather than a full-fledged MTA. So far I've managed to work around this limitation with custom scripts, but I've recently realized a need to include Variable Envelope Return Paths in order to help process incoming emails that hardfail. These are typically auto-replies or bounces for which Postal cannot find a route. Is this a direction the Postal team would want to take? It seems there are a few discussions previously with similar issues which may benefit from such a solution (see #2705, #1614, #1097, #1298, etc.)

Without looking too deeply into this so far, it seems that this change would be a minimal impact to existing code, namely lines 144 and 146 in 
 /lib/postal/smtp_sender.rb to include either the recipient's address or the message id from the original message headers:

https://github.com/postalserver/postal/blob/18ba7140b4dec603569bef8d61b869c5f4404766/lib/postal/smtp_sender.rb#L143C1-L146C81

        elsif message.domain.return_path_status == "OK"
          mail_from = "#{message.server.token}@#{message.domain.return_path_domain}"
        else
          mail_from = "#{message.server.token}@#{Postal.config.dns.return_path}"

I should be able to write this code, although I'm unfamiliar with Ruby. If this isn't a direction Postal would like to take as a pull request, would I be able to fork Postal?

**Comment by willpower232** [comment]
Its highly unlikely the maintainers would want to drift from the transactional email too far but I feel like people have asked for this in the past.

Fundamentally, Postal only really works if it knows about bounces which is why this probably wouldn't change. Also there could be issues with validation of the message if it doesn't line up properly so that might also be a thing.

Postal is open source so you should be able to fork it and do what you want with it but you would have to create your own docker images and probably change the install script to match them.

**Comment by jmdunsing** [comment]
I submitted PR #2855 with a coding change in case the maintainers want to incorporate this feature.

**Comment by adamcooke** [comment]
I don't have any objection to this in principle. However, this would also need to change the way that bounces are attributed to the outgoing messages when they are de-queued.

Thinking about it now, I don't really know why we didn't do this from the start. There may or may not be a good reason for this.
